examples/src/test/scala/GithubExample.scala

```diff
@@ -0,0 +1,310 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+import cats._
+import cats.data.NonEmptyList
+import cats.effect._
+import cats.instances.list._
+import cats.syntax.all._
+import cats.temp.par._
+
+import io.circe._
+import io.circe.generic.semiauto._
+
+import org.http4s._
+import org.http4s.headers._
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.circe._
+import org.http4s.client._
+import org.http4s.client.dsl._
+import org.http4s.client.blaze._
+import org.scalatest.{Matchers, WordSpec}
+
+import fetch.{DataSource, Env, Fetch}
+
+class GithubExample extends WordSpec with Matchers {
+  implicit val executionContext = ExecutionContext.Implicits.global
+
+  val ACCESS_TOKEN: String = sys.env("ORG_GITHUB_TOKEN")
+
+  implicit val t: Timer[IO]         = IO.timer(executionContext)
+  implicit val cs: ContextShift[IO] = IO.contextShift(executionContext)
+
+  // http4s client which is used by the datasources
+
+  def client[F[_]: ConcurrentEffect] =
+    Http1Client[F](BlazeClientConfig.defaultConfig)
+
+  // Github HTTP api
+
+  val GITHUB: Uri = Uri.unsafeFromString("https://api.github.com")
+
+  val REL_NEXT = "rel=\"next\"".r
+
+  def hasNext[F[_]: ConcurrentEffect](res: Response[F]): Boolean =
+    res.headers
+      .get(CaseInsensitiveString("Link"))
+      .fold(false)({ h =>
+        REL_NEXT.findFirstIn(h.value).isDefined
+      })
+
+  def getNextLink[F[_]: ConcurrentEffect](raw: String): F[String] = {
+    REL_NEXT
+      .findFirstMatchIn(raw)
+      .fold(
+        Sync[F].raiseError[String](new Exception("Couldn't find next link"))
+      )(m => {
+        Sync[F].pure(m.before.toString.split(",").last.trim.dropWhile(_ == '<').takeWhile(_ != '>'))
+      })
+  }
+
+  def getNext[F[_]: ConcurrentEffect](res: Response[F]): F[Uri] =
+    res.headers
+      .get(CaseInsensitiveString("Link"))
+      .fold(Sync[F].raiseError[Uri](new Exception("next not found")))(
+        raw => getNextLink(raw.value).map(Uri.unsafeFromString(_))
+      )
+
+  // -- repos
+
+  type Org = String
+  case class Repo(
+      name: String,
+      fork: Boolean,
+      forks_count: Int,
+      stargazers_count: Int,
+      watchers_count: Int,
+      languages_url: String,
+      contributors_url: String
+  )
+
+  implicit val repoD: Decoder[Repo] = deriveDecoder
+
+  private def fetchOrgRepos[F[_]](c: Client[F], req: Request[F])(
```

Comment:
These three functions follow the same pattern, I'm guessing if we could just write one generic function and then move the `hasNext`, `getNextLink` and `getNext` inside of it:
```scala
private def fetchDataRecursively[F[_]: ConcurrentEffect, A](c: Client[F], req: Request[F])(decode: Response[F] => F[List[A]])(
      implicit C: ConcurrentEffect[F]
  ): F[List[A]] = {

    val REL_NEXT = "rel=\"next\"".r

    def hasNext(res: Response[F]): Boolean =
      res.headers
        .get(CaseInsensitiveString("Link"))
        .fold(false)({ h =>
          REL_NEXT.findFirstIn(h.value).isDefined
        })

    def getNextLink(raw: String): F[String] = {
      REL_NEXT
        .findFirstMatchIn(raw)
        .fold(
          Sync[F].raiseError[String](new Exception("Couldn't find next link"))
        )(m => {
          Sync[F].pure(m.before.toString.split(",").last.trim.dropWhile(_ == '<').takeWhile(_ != '>'))
        })
    }

    def getNext(res: Response[F]): F[Uri] =
      res.headers
        .get(CaseInsensitiveString("Link"))
        .fold(Sync[F].raiseError[Uri](new Exception("next not found")))(
          raw => getNextLink(raw.value).map(Uri.unsafeFromString(_))
        )

    for {
      result <- c.fetch[List[A]](req) {
        case Status.Ok(res) =>
          if (hasNext(res)) {
            for {
              data <- decode(res)
              nxt  <- getNext(res)
              newReq = req.withUri(nxt)
              moreData <- fetchDataRecursively(c, newReq)(decode)
            } yield data ++ moreData
          } else
            res.as[List[A]]
        case res =>
          C.raiseError(new Exception(res.body.toString))
      }
    } yield result
  }
```

Reply:
Then we could have:
```scala
private def fetchOrgRepos[F[_]](c: Client[F], req: Request[F])(
      implicit C: ConcurrentEffect[F]
  ): F[List[Repo]] = {
    implicit val reposED: EntityDecoder[F, List[Repo]] = jsonOf
    fetchDataRecursively[F, Repo](c, req)(_.as[List[Repo]])
  }

private def fetchLanguages[F[_]](c: Client[F], req: Request[F])(
      implicit C: ConcurrentEffect[F]
  ): F[List[Language]] = {
    implicit val objED: EntityDecoder[F, JsonObject] = jsonOf
    fetchDataRecursively[F, Language](c, req)(_.as[JsonObject].map(_.toList.map(_._1)))
  }

private def fetchContributors[F[_]](c: Client[F], req: Request[F])(
      implicit C: ConcurrentEffect[F]
  ): F[List[Contributor]] = {
    implicit val objED: EntityDecoder[F, List[Contributor]] = jsonOf
    fetchDataRecursively[F, Contributor](c, req)(_.as[List[Contributor]])
  }
```